internal/workload/workload.go

```diff
@@ -195,5 +225,83 @@ func (w *WorkloadManager) toPodYaml(workload *models.Workload) ([]byte, error) {
 	}
 	pod.Kind = "Pod"
 	pod.Name = workload.Name
-	return yaml.Marshal(pod)
+	return &pod, nil
+}
+
+func (ww workloadWrapper) Init() error {
+	return ww.netfilter.AddTable(nftTableName)
+}
+
+func (ww workloadWrapper) List() ([]api2.WorkloadInfo, error) {
+	return ww.workloads.List()
+}
+
+func (ww workloadWrapper) Remove(workloadName string) error {
+	if err := ww.workloads.Remove(workloadName); err != nil {
+		return err
+	}
+	if err := ww.netfilter.DeleteChain(nftTableName, workloadName); err != nil {
+		log.Errorf("Failed to delete chain %[1]s from %s table for workload %[1]s", workloadName, nftTableName)
+	}
+	return nil
+}
+
+func (ww workloadWrapper) Run(workload *v1.Pod, manifestPath string) error {
+	if err := ww.applyNetworkConfiguration(workload); err != nil {
+		return err
+	}
+	if err := ww.workloads.Run(manifestPath); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ww workloadWrapper) applyNetworkConfiguration(workload *v1.Pod) error {
+	hostPorts, err := getHostPorts(workload)
+	if err != nil {
+		return err
+	}
+	if len(hostPorts) == 0 {
+		return nil
+	}
+	//TODO: should verify if chain doesn't exist
+	if err := ww.netfilter.AddChain(nftTableName, workload.Name); err != nil {
+		// should we return nil and only log the error ?
```

Comment:
I think that network configuration is part of the workload. We need to validate this in the operator before attempting to deploy and here we fail and should not continue with the deployment. I think it is good to log here as well. It will help to understand why workload deployment failed.

Reply:
iinw, the pod spec validates the correctness of the ports range, but not sure if we can do anything more than that on the operator side (as we don't expose any network configuration, at least not at this stage).
I'll return the error from this context and log it from its caller.

Reply:
I think what we can check is port collisions (if already used by other workload). Is there anything else which can cause `AddChain` to return an error?

Reply:
the port collisions will not be detected by nft. you can have the same rule in more than one chain. e.g.:
```
table inet edge {
        chain nginx-sample-x {
                type filter hook input priority filter; policy accept;
                tcp dport 10000 ct state established,new counter packets 0 bytes 0 accept
        }

        chain nginx-sample1 {
                type filter hook input priority filter; policy accept;
                tcp dport 9980 ct state established,new counter packets 0 bytes 0 accept
        }

        chain nginx-sample2 {
                type filter hook input priority filter; policy accept;
                tcp dport 10000 ct state established,new counter packets 0 bytes 0 accept
        }
}
```

ports collision will be detected by podman. also, there is a chance other service might occupy the port (perhaps not managed by k4e operator - if such scenario makes any sense).
I'd suggest to address failures when running pods by podman, and in case of a failure to do clear any resources resources related to the workload. On next attempt to run the workload, those resources will be recreated.

On the management level, ports collision can be validated on cluster side, when adding or updating edgedeployment resources of the same edge device, and deny attempts to run edgedeployment  if a port is already taken by other edgedeployment.

Reply:
the port collisions will not be detected by nft. you can have the same rule in more than one chain. e.g.:
```
table inet edge {
        chain nginx-sample-x {
                type filter hook input priority filter; policy accept;
                tcp dport 10000 ct state established,new counter packets 0 bytes 0 accept
        }

        chain nginx-sample1 {
                type filter hook input priority filter; policy accept;
                tcp dport 9980 ct state established,new counter packets 0 bytes 0 accept
        }

        chain nginx-sample2 {
                type filter hook input priority filter; policy accept;
                tcp dport 10000 ct state established,new counter packets 0 bytes 0 accept
        }
}
```

ports collision will be detected by podman. also, there is a chance other service might occupy the port (perhaps not managed by k4e operator - if such scenario makes any sense).
I'd suggest to address failures when running pods by podman, and in case of a failure to do clear any resources resources related to the workload. On next attempt to run the workload, those resources will be recreated.

On the management level, ports collision can be validated on cluster side, when adding or updating edgedeployment resources of the same edge device, and deny attempts to run edgedeployment  if a port is already taken by other edgedeployment.